### ISSUE TITLE
RNN ++

### ISSUE THREAD

**Issue by nicholas-leonard** [nicholas-leonard/dp#100]
recurrentlanguagemodel.lua isn't converging bellow 320 PPL. Possible causes:
- Learning rate isn't being decayed
- RNN shouldn't forget previous context after BPTT update (I was told that it should, but still not convinced)
- Max-norm or weight decay regularization (too high or too low)
- SoftmaxTree (yet it works for languagemodel.lua)

Possible optimizations:
- acc_update the output layer to save memory (will cause instability in the gradients for BPTT)
